query/src/storages/system/engines_table.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::any::Any;
+use std::sync::Arc;
+
+use common_datablocks::DataBlock;
+use common_datavalues::prelude::*;
+use common_exception::Result;
+use common_meta_types::TableIdent;
+use common_meta_types::TableInfo;
+use common_meta_types::TableMeta;
+use common_planners::ReadDataSourcePlan;
+use common_streams::DataBlockStream;
+use common_streams::SendableDataBlockStream;
+
+use crate::catalogs::Catalog;
+use crate::sessions::QueryContext;
+use crate::storages::Table;
+
+pub struct EnginesTable {
+    table_info: TableInfo,
+}
+
+impl EnginesTable {
+    pub fn create(table_id: u64) -> Self {
+        let schema = DataSchemaRefExt::create(vec![
+            DataField::new("Engine", DataType::String, false),
+            DataField::new("Comment", DataType::String, false),
+        ]);
+
+        let table_info = TableInfo {
+            desc: "'system'.'engines'".to_string(),
+            name: "engines".to_string(),
+            ident: TableIdent::new(table_id, 0),
+            meta: TableMeta {
+                schema,
+                engine: "SystemEngines".to_string(),
+                ..Default::default()
+            },
+        };
+        EnginesTable { table_info }
+    }
+}
+
+#[async_trait::async_trait]
+impl Table for EnginesTable {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn get_table_info(&self) -> &TableInfo {
+        &self.table_info
+    }
+
+    async fn read(
+        &self,
+        _ctx: Arc<QueryContext>,
+        _plan: &ReadDataSourcePlan,
+    ) -> Result<SendableDataBlockStream> {
+        let descriptors = _ctx.get_catalog().get_storage_descriptors();
```

Comment:
Why we add `StorageDescriptor` to catalog and impl it for every table engine, how about get the table engine name and descriptor from self.ctx.storage_factory?
It seems we only add a function like `desc()->&str` for the table engine.

Reply:
i think i don't get the real meaning of this. can you explain it more clearly? i think every engine should has it's own creator

Reply:
the reason of why we add StorageDescriptor to catalog is that storage_factory is only MutableCatalog's member variable，so database catalog should expose a api to get descriptors, or we can get descriptors from storage_factory and construct the EnginesTable

Reply:
Descrptor is Attributes of engine，so I impl it for every table engine...

Reply:
How about only add one function like description() to table engine instead of the Descriptior trait? And we register the description to storage factory, this looks more clearly.
Then we can add one trait function named 'get_table_engines' to Catalog, which getting the engine details from the mutable catalog.